lib/std/unicode.zig

```diff
@@ -231,6 +231,128 @@ pub fn utf8ValidateSlice(input: []const u8) bool {
+            four3,
+            four4,
+        };
+
+        pub fn byte_class(byte: u8) ByteClass {
+            return switch (byte) {
+                0x00...0x7f => .ascii,
+                0x80...0x8f => .cont1,
+                0x90...0x9f => .cont2,
+                0xa0...0xbf => .cont3,
+                0xc0...0xc1 => .two1,
+                0xc2...0xdf => .two2,
+                0xe0...0xe0 => .three1,
+                0xe1...0xec => .three2,
+                0xed...0xed => .three3,
+                0xee...0xef => .three2,
+                0xf0...0xf0 => .four1,
+                0xf1...0xf3 => .four2,
+                0xf4...0xf4 => .four3,
+                0xf5...0xff => .four4,
+            };
+        }
+
+        const State = enum(u5) {
+            ok = 8,
+            one = 13,
+            two1 = 23,
+            two2 = 18,
+            three1 = 3,
+            three2 = 28,
+            fail = 0,
+        };
+
+        const start: State = .ok;
+        const accept: []State = .{.ok};
+        const fail: ?State = .fail;
+
+        fn step(byte: u8, state: State) State {
+            const class = byte_class(byte);
+            return switch (state) {
+                .ok => switch (class) {
+                    .ascii => .ok,
+                    .cont1 => .one,
+                    .cont2 => .one,
+                    .cont3 => .one,
+                    else => .fail,
+                },
+                .one => switch (class) {
+                    .two2 => .ok,
+                    .cont1 => .two1,
+                    .cont2 => .two1,
+                    .cont3 => .two2,
+                    else => .fail,
+                },
+                .two1 => switch (class) {
+                    .three2 => .ok,
+                    .three3 => .ok,
+                    .cont1 => .three1,
+                    .cont2 => .three2,
+                    .cont3 => .three2,
+                    else => .fail,
+                },
+                .two2 => switch (class) {
+                    .three1 => .ok,
+                    .three2 => .ok,
+                    .three3 => switch (surrogates) {
+                        .cannot_encode_surrogate_half => .fail,
+                        .can_encode_surrogate_half => .ok,
+                    },
+                    .cont1 => .three1,
+                    .cont2 => .three2,
+                    .cont3 => .three2,
+                    else => .fail,
+                },
+                .three1 => switch (class) {
+                    .four2 => .ok,
+                    .four3 => .ok,
+                    else => .fail,
+                },
+                .three2 => switch (class) {
+                    .four1 => .ok,
+                    .four2 => .ok,
+                    else => .fail,
+                },
+                .fail => .fail,
+            };
+        }
+
+        const shift_table = blk: {
+            @setEvalBranchQuota(30000);
+            var t = [_]u32{0} ** 256;
```

Comment:
This is a total nitpick, but why not just use `@splat` to initialize the shift table? At least for me, this makes the type of the shift table quicker to understand (even if it only takes a few seconds to understand it to begin with), and means the compiler doesn't have to do any comptime list concatenation.
```suggestion
            var t: [256]u32 = @splat(0);
```